src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -1539,3 +1545,105 @@ SyncMetadataToNodes(void)
 
 	return result;
 }
+
+
+/*
+ * SyncMetadataToNodesMain is the main function for syncing metadata to
+ * MX nodes. It retries until success and then exits.
+ */
+int
+SyncMetadataToNodesMain(Datum main_arg)
+{
+	Oid databaseOid = DatumGetObjectId(main_arg);
+
+	/* extension owner is passed via bgw_extra */
+	Oid extensionOwner = InvalidOid;
+	memcpy_s(&extensionOwner, sizeof(extensionOwner),
+			 MyBgworkerEntry->bgw_extra, sizeof(Oid));
+
+	/* connect to database, after that we can actually access catalogs */
+	BackgroundWorkerInitializeConnectionByOid(databaseOid, extensionOwner, 0);
+
+	/* make worker recognizable in pg_stat_activity */
+	pgstat_report_appname("Citus Metadata Sync Daemon");
+
+	bool syncedAllNodes = false;
+
+	InvalidateMetadataSystemCache();
+	StartTransactionCommand();
+
+	/*
+	 * Some functions in ruleutils.c, which we use to get the DDL for
+	 * metadata propagation, require an active snapshot.
+	 */
+	PushActiveSnapshot(GetTransactionSnapshot());
+
+	if (!LockCitusExtension())
+	{
+		ereport(DEBUG1, (errmsg("could not lock the citus extension, "
+								"skipping metadata sync")));
+	}
+	else if (CheckCitusVersion(DEBUG1) && CitusHasBeenLoaded())
+	{
+		MetadataSyncResult result = SyncMetadataToNodes();
+
+		syncedAllNodes = (result == METADATA_SYNC_SUCCESS);
+
+		/*
+		 * Notification means we had an attempt on synchronization
+		 * without being blocked for pg_dist_node access.
+		 */
+		if (result != METADATA_SYNC_FAILED_LOCK)
+		{
+			Async_Notify(METADATA_SYNC_CHANNEL, NULL);
+		}
+	}
+
+
+	PopActiveSnapshot();
+	CommitTransactionCommand();
+	ProcessCompletedNotifies();
+
+	if (syncedAllNodes)
+	{
+		return 0;
+	}
+
+	return 1;
+}
+
+
+/*
+ * SpawnSyncMetadataToNodes starts a background worker which runs metadata
+ * sync. On success it returns workers' handle. Otherwise it returns NULL.
+ */
+BackgroundWorkerHandle *
+SpawnSyncMetadataToNodes(Oid database, Oid extensionOwner)
+{
+	BackgroundWorker worker;
+	BackgroundWorkerHandle *handle = NULL;
+
+	/* Configure a worker. */
+	memset(&worker, 0, sizeof(worker));
+	SafeSnprintf(worker.bgw_name, sizeof(worker.bgw_name),
+				 "Citus Metadata Sync: %u/%u",
+				 database, extensionOwner);
+	worker.bgw_flags =
+		BGWORKER_SHMEM_ACCESS | BGWORKER_BACKEND_DATABASE_CONNECTION;
+	worker.bgw_start_time = BgWorkerStart_ConsistentState;
+	worker.bgw_restart_time = (MetadataSyncRetryInterval + 999) / 1000;
```

Comment:
what's the +999?

Reply:
This is to get the ceiling of division by 1000 instead of its floor. bgw_restart_time is integer seconds while MetadataSyncRetryInterval is in milliseconds.

Alternatively we could:
1. wait in precise milliseconds by implementing a loop inside the background worker's main, which might make the code a bit more complicated.
2. change units of MetadataSyncRetryInterval to seconds

I will try to implement option 2.